### DISCUSSION TITLE
How can I manipulate the way conditional probability table is formatted?

### DISCUSSION THREAD

**Discussion by kenneth-lee-ch** [agrumery/aGrUM#5] [Q&A] [answered]
Suppose I have some random BN generated and I generate some conditional probability table using `cpt()`. Given a string, how can I always be able to get a table where the given string is on the column instead of rows?

For example, here I attached an example, let's say I want to have B to be where D is always for any given bn.cpt() whenever B exists in the cpt. 

Is there a way to do that for any given cpt?

BEFORE:
![Screenshot 2023-02-15 at 5 12 37 AM](https://user-images.githubusercontent.com/24616330/218998625-b2758dc4-66e5-4cf7-9e10-10bc66a3459b.png)

AFTER:
![Screenshot 2023-02-15 at 5 17 44 AM](https://user-images.githubusercontent.com/24616330/218999759-3c4e0f77-0c7a-40fe-98ec-2d5d4cc656f7.png)

**Comment by phwuil** [comment]
Hi @kenneth-lee-ch ,

Is this what you want to do ? (I do not really understand what you mean by"given a string")

![image](https://user-images.githubusercontent.com/5980636/219180136-bd91f083-5569-41ab-8792-600aad77c37d.png)

Note that when the Potential is on a white background, it means that it is not a CPT anymore :-)

**Comment by kenneth-lee-ch** [comment]
@phwuil  I am just thinking about how to get the conditional probability given any target and conditioning set from a CPT and be able to output a prediction according to the largest probability given a sample. I was thinking to get `getPosterior`, but it doesn't seem like the same.

**Reply by phwuil** [reply]
Ok, if I understand correctly this time, you want to use : `LazyPropagation.evidenceIImpact` :
﻿
<img width="419" alt="image" src="https://user-images.githubusercontent.com/5980636/219856169-184f5b67-b3b6-4630-ad83-cc4eca9dfa9e.png">

Note that  `evidenceImpact` above returns P(C|A,F) since E is d-separated from C given A and F => P(C|A,E,F)=P(C|A,F)

**Reply by kenneth-lee-ch** [reply]
How is this different from `getPosterior()`?

**Comment by phwuil** [comment, marked as answer]
It is a `getPosterior` for all the possible configurations of  the variables in the evidence. 

`getPosterior` : P(X |Y=1,Z=0)
`evidenceImpact` : the whole P(X|Y=y,Z=z) forall y,z

Moreover, it performs this d-separation analysis... 

And it takes advantage of aGrUM's ability to make incremental inferences: some computations can be factored. Such that an `evidenceImpact` if faster than several `getPosterior`s.

**Comment by kenneth-lee-ch** [comment]
@phwuil  I have a follow-up question. If I use `causalImpact` to get the post-interventional distribution from the returned argument `computation`. How can I manipulate the probability table such that the one shown on the screenshot such that the column corresponding to a specific variable I look for e.g. C instead of L?  In other words, swapping the column and a specific row according to a specific variable name? I don't know how to do that for any arbitrary table output by `causalImpact`.

The reason why I want to do that is that I want to get a list of values (0 and 1) according to the max value of a specific variable in the table. For example, in the screenshot, if my specific variable is W, then I would get all 0's as the list of values because 0 always corresponds to the higher probability in this case. 

![Screenshot 2023-02-28 at 9 15 02 PM](https://user-images.githubusercontent.com/24616330/222027612-938503bf-48e2-4ea3-bfb9-5c4cbce7b2b4.png)

**Comment by phwuil** [comment]
Hi @kenneth-lee-ch , you have two methods of Potential to do so : `p.putFirst(varname)` and `p.reorganize(list_of_varnames)` :

<img width="880" alt="image" src="https://user-images.githubusercontent.com/5980636/222078055-26422456-173d-4333-b474-0216aa379f45.png">

By the way, if I well understood what you would like :

<img width="432" alt="image" src="https://user-images.githubusercontent.com/5980636/222078453-acae2765-9e77-407d-9af7-ec268bccd865.png">

**Reply by kenneth-lee-ch** [reply]
@phwuil If I have a dataframe with binary variables `F, C, L` as columns, let's say I want to treat entries of your `bestW` potential as the predicted value (e.g. when F=0, C=0, L=0, my predicted value is 0; when F=1, C=0, L=0, my predicted value is 1), what's the efficient way to fill in the dataframe with these predicted values accordingly?

**Reply by phwuil** [reply]
`p.topandas()`?

![image](https://user-images.githubusercontent.com/5980636/222946118-dcd8a363-5b95-4fae-975c-207caa8b257b.png)

**Reply by kenneth-lee-ch** [reply]
@phwuil  No, that's not what I meant. Suppose you have a dataframe with columns `E,C,L` and they are all binary variables. Now, predict the value of W based on the potential CPT with the entries of `E,C,L` you have from the dataframe. Let's say the first row of the dataframe tells you `L=0, C=0, E= 0`, then you will predict `W=0` based on the largest probability between `W=0` vs `W=1`

**Reply by phwuil** [reply]
Given the `bestW` Potential above, it is pretty straightforward :-)

Just a tip : `df.to_dict(orient='records')`

... and then : `[(d,bestW[d] ) for d in df.to_dict(orient='records')]`

**Reply by kenneth-lee-ch** [reply]
@phwuil I actually had a way before I asked the question. Just wanted to see if there is a more efficient way that can be done through the package functions. Thanks though.
